lib/util/traffic-switch/traffic-switcher.ts

```diff
@@ -0,0 +1,139 @@
+    }
+
+    this.experimentName = experimentName
+    this.implementationA = implementationA
+    this.implementationB = implementationB
+    this.aliasA = aliasA
+    this.aliasB = aliasB
+    this.pctTrafficSwitchToB = pctTrafficSwitchToB
+    this.pctSamplingToCompare = pctSamplingToCompare
+  }
+
+  /* This method will do one of the following:
+   *  - Switch traffic between two implementations based on the pctTrafficSwitchToB percentage
+   *  - Sample traffic between two implementations based on the pctSamplingToCompare percentage
+   */
+  protected async trafficSwitchMethod<K>(
+    methodA: () => Promise<K>,
+    methodB: () => Promise<K>,
+    methodName: string,
+    defaultReturnVal: K,
+    comparisonMethod: ((resultA: K | undefined, resultB: K | undefined) => void) | undefined
+  ): Promise<K> {
+    if (this.isSamplingEnabled()) {
+      if (this.shouldSample()) {
+        this.logMetric(methodName, `COMPARISON_SAMPLE`)
+        // If in sampling mode, and we should sample, call both implementations and compare results
+        // Note: here we'll wait for both implementations to finish before comparing results.
+        const results: (K | undefined)[] = await this.allSettled<K>([
+          [methodA(), this.aliasA, methodName],
+          [methodB(), this.aliasB, methodName],
+        ])
+
+        // Compare the results if a comparison method is provided
+        if (results && results.length == 2) {
+          try {
+            if (comparisonMethod) {
+              comparisonMethod(results[0], results[1])
+            }
+          } catch (error) {
+            log.error(`Error in TrafficSwitcher comparison method ${methodName}`, error)
+            this.logMetric(methodName, 'COMPARISON_ERROR')
+          }
+        }
+
+        // Always return the result from the first implementation when in sampling mode, or default value.
+        // Note: when in sampling mode, exceptions in sampling traffic are logged and ignored (default value is returned).
+        return results && results.length > 0 ? results[0] ?? defaultReturnVal : defaultReturnVal
+      } else {
+        // If in sampling mode, but we are not sampling for this call, just call implementation A.
+        return methodA()
+      }
+    } else {
+      // Otherwise we are in switch traffic mode, just call the selected implementation based on the traffic switch percentage
+      if (Math.random() < this.pctTrafficSwitchToB) {
+        this.logMetric(methodName, `SELECTED_IMPL__${this.aliasB}`)
+        return methodB()
+      } else {
+        this.logMetric(methodName, `SELECTED_IMPL__${this.aliasA}`)
+        return methodA()
+      }
+    }
+  }
+
+  protected isSamplingEnabled(): boolean {
+    return this.pctSamplingToCompare > 0
+  }
+
+  protected shouldSample(): boolean {
+    return this.isSamplingEnabled() && Math.random() < this.pctSamplingToCompare
+  }
```

Comment:
This will end up sampling on every method call rather than throughout the life of the request. This is okay for this use-case, but in a real experiment we will want to do this sampling in a more deterministic way, perhaps by hashing the requestId?

Reply:
good call - currently logic is to sample/traffic switch at the per call/method level (`trafficSwitchMethod`) 
I agree probably in the future we'll need something stickier and requestId is a good candidate.
I will add a note + TODO for this, as it might require a bit more thinking to get it right.

Reply:
I see `TrafficSwitcher` as a work in progress class that will get more complete over time (e.g. your idea, plus we could add timing comparison etc).

Reply:
yup yup, I agree, just doing some brainstorming here about the directions we could take this, but leaving them as good to haves for now